server/client.go

```diff
@@ -2050,6 +2079,40 @@ func (c *client) queueOutbound(data []byte) {
 	// Add to pending bytes total.
 	c.out.pb += int64(len(data))
 
+	// Take a copy of the slice ref so that we can chop bits off the beginning
+	// without affecting the original "data" slice.
+	toBuffer := data
+
+	// All of the queued []byte have a fixed capacity, so if there's a []byte
+	// at the tail of the buffer list that isn't full yet, we should top that
+	// up first. This helps to ensure we aren't pulling more []bytes from the
+	// pool than we need to.
+	if len(c.out.nb) > 0 {
+		last := &c.out.nb[len(c.out.nb)-1]
+		if free := cap(*last) - len(*last); free > 0 {
+			if l := len(toBuffer); l < free {
+				free = l
+			}
+			*last = append(*last, toBuffer[:free]...)
+			toBuffer = toBuffer[free:]
+		}
+	}
+
+	// Now we can push the rest of the data into new []bytes from the pool
+	// in fixed size chunks. This ensures we don't go over the capacity of any
+	// of the buffers and end up reallocating.
+	for len(toBuffer) > 0 {
+		new := nbPoolGet(len(toBuffer))
```

Comment:
If we have existing nb, we should just do another 64k and not fragment. Should only be small iff no other pending imo.

Reply:
By this point we've already "filled up" the existing last buffer in `nb` and we will only allocate new buffers if there's still something we've yet to queue up. That way we can reduce the number of vectors to `writev` for lots of small successive messages.

Reply:
The small buffers should only be used if no pending at all imo, for the ping to all issue. If there is any pending they should all be 64k. Since that means the publisher is publishing at a fast rate.

Reply:
I can modify it to ensure that later buffers are always large ones, sure.

Reply:
Yes, so if the message to be queued is small, and if there is no pending get a small buffer if the message will fit. Otherwise always do 64k.